### PR TITLE
Fix handling of partial packets

### PR DESCRIPTION
**Pull request by codido**
BufReader::fill_buf() only fills more data from the inner reader if the
internal buffer is empty. If a partial packet is read, calling
fill_buf() in process_packets_from() again wont read the rest of the
packet, resulting in an endless loop.

In addition, the PacketSize feature is parsed as a hex number by gdb, so
returning 65536 is actually interpreted as 0x65536.

### CODE REVIEW

**Comment by luser** [comment]
This might be a bit easier to read (and less fragile for future changes) if you pulled in the [bytes crate](https://crates.io/crates/bytes) and used a [`BytesMut`](https://docs.rs/bytes/0.5.2/bytes/struct.BytesMut.html) as the buffer. You'd still need an additional array or `Vec` to pass directly to `read`, but then when you got new data you'd pass it to [`extend_from_slice`](https://docs.rs/bytes/0.5.2/bytes/struct.BytesMut.html#method.extend_from_slice), then pass the whole `BytesMut` to `run_parser`, then call [`split_to`](https://docs.rs/bytes/0.5.2/bytes/struct.BytesMut.html#method.split_to) to remove however many bytes were parsed from the buffer.

**Comment by codido** [comment]
@luser thanks for the prompt review!

The first version was actually similar to what you suggested (with an extra vector instead of bytes, but similar approach), but wanted to keep changes minimal and avoid needless copying. Granted, this probably wont be a performance bottleneck though :)

I don't mind making such changes, though not sure it'll be more readable than this patch. And since we there's no need to support arbitrary sized packets anyway, I'm not sure it's worth while?

**Comment by codido** [comment]
By the way, wouldn't using bytes in this manner involve more heap operations needlessly? Though, again, performance is probably not a huge concern here..

**Comment by luser** [comment]
> By the way, wouldn't using bytes in this manner involve more heap operations needlessly? Though, again, performance is probably not a huge concern here..

Yeah, you're correct. I'm just squeamish about manual byte position tracking like this but I'm probably overthinking it. :)

File: src/lib.rs [hunk 1]
```diff
@@ -1424,47 +1424,40 @@ fn run_parser(buf: &[u8]) -> Option<(usize, Packet)> {
...
-pub fn process_packets_from<R, W, H>(reader: R, mut writer: W, handler: H)
+pub fn process_packets_from<R, W, H>(mut reader: R, mut writer: W, handler: H)
 where
     R: Read,
     W: Write,
     H: Handler,
 {
-    let mut bufreader = BufReader::with_capacity(MAX_PACKET_SIZE, reader);
-    let mut done = false;
+    let mut buf = [0; MAX_PACKET_SIZE];
```
Comment (luser) [inline src/lib.rs]: 64K is pretty big for a stack buffer, I think this ought to be a `Vec`.

**Comment by luser** [comment]
I think with that one change this should be fine to merge!

**Comment by codido** [comment]
Sure thing, amended the commit, thanks!

**Comment by luser** [comment]
Thanks for the fix!

**Pull request merged**